### PR TITLE
Create v2 improved 

### PR DESCRIPTION
**Pull request by 5oappy**
## Describe your changes
improved create function by following testing criteria from John

image

Code has been written and briefly tested to meet all criteria except 'weird periodicity' as team is still in the process of discussing how to handle it. userId verification has also been left untouched as I believe that is server side and is not crucial for this sprint's scope. Furthermore 'merge (time) interval' appears to be more related to the unavailability update method so I would like to handle that separately to this pr.

Right now the test cases I have manually tested in postman that have resulted in expected behaviour consist of:

- no overlapping events
- overlapping start time
- overlapping end time
- fully enclosed event, new event that encloses an editing event
- existing fully enclosed, new event inside an existing event
- overlapping periodicity (still unsure as we need to further establish the periodicity schema)
- no overlapping periodicity 
- exact match, duplicate
- new event starts exactly on end time of exisiting event

I have not found any unexpected behaviour, however, keep in mind that it has only been manually tested with a couple of examples each so may have missed something.

## Issue ticket number and link
https://fireapp-emergiq-2024.atlassian.net/browse/FE-68

### CODE REVIEW

**Comment by 5oappy** [comment]
I'm actually not sure what's wrong with the naming as it worked last time, please let me know how to fix when available :)

**Comment by TaiHaDev** [comment]
I think merge time interval functionality should be used for both create and update so we should write a repository function for that. Do you mind writing it @5oappy?

**Comment by 5oappy** [comment]
> I think merge time interval functionality should be used for both create and update so we should write a repository function for that. Do you mind writing it @5oappy?

I'm not sure if it'd work for this sprint but definitely before next sprint begins ill have something done for ready to be reviewed.

File: controllers/v2/unavailability/api.py [hunk 1]
```diff
@@ -50,7 +55,7 @@ class VolunteerUnavailabilityV2(Resource):
 
     @requires_auth
     @marshal_with(volunteer_unavailability_time)
-    @is_user_or_has_role(None, UserType.ROOT_ADMIN)
+    # @is_user_or_has_role(None, UserType.ROOT_ADMIN)
```
Comment (BenMMcLean) [inline controllers/v2/unavailability/api.py:58]: This should be secure, revert this change

File: controllers/v2/unavailability/api.py [hunk 2]
```diff
@@ -64,7 +69,26 @@ def get(self, user_id):
     def post(self, user_id):
         try:
             args = edit_parser.parse_args()
+            # Check if start time is earlier than end time
+            if args['start'] >= args['end']:
+                return {"description": "Start time must be earlier than end time"}, 400  # HTTP 400 Bad Request
```
Comment (BenMMcLean) [inline controllers/v2/unavailability/api.py:74]: "message" rather than "description" is more in line with the rest of the app (please do this to the bit below as well)

File: controllers/v2/unavailability/api.py [hunk 3]
```diff
@@ -64,7 +69,26 @@ def get(self, user_id):
...
+                    UnavailabilityTime.end > args['start'],
+                    UnavailabilityTime.periodicity == args['periodicity']
+                ).all()
+                if overlapping_events:
+                    overlapping_details = []
+                    for event in overlapping_events:
+                        overlapping_details.append({
+                            "eventId": event.eventId})
+                    return {"description": "Time frames overlap with existing events",
+                            "overlappingEvents": overlapping_details}, 400  # HTTP 400 Bad Request
```
Comment (BenMMcLean) [inline controllers/v2/unavailability/api.py:90]: Please use snake_case

**Review by BenMMcLean — CHANGES_REQUESTED** [review]
Generally happy with your changes, but your branch name needs to be updated to match our conventions, and there are a few minor fixes

File: controllers/v2/unavailability/api.py (hunk 1 above)
Comment (5oappy) [reply controllers/v2/unavailability/api.py:58]: huh thats weird I swear I reverted that already will do
